--- a/arch/arm/dts/Makefile	2019-01-15 00:02:36.000000000 +0200
+++ b/arch/arm/dts/Makefile	2019-05-03 17:50:22.000000000 +0300
@@ -376,6 +376,7 @@
 	sun8i-h2-plus-libretech-all-h3-cc.dtb \
 	sun8i-h2-plus-orangepi-r1.dtb \
 	sun8i-h2-plus-orangepi-zero.dtb \
+	sun8i-h2-plus-orangepi-zero-lts.dtb \
 	sun8i-h3-bananapi-m2-plus.dtb \
 	sun8i-h3-libretech-all-h3-cc.dtb \
 	sun8i-h3-nanopi-m1.dtb \
--- a/arch/arm/dts/sun8i-h2-plus-orangepi-zero-lts.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/dts/sun8i-h2-plus-orangepi-zero-lts.dts	2019-05-03 18:15:54.000000000 +0300
@@ -0,0 +1,9 @@
+
+/* Orange Pi Zero Lts is based on Orange Pi Zero design */
+#include "sun8i-h2-plus-orangepi-zero.dts"
+
+/ {
+	model = "Xunlong Orange Pi Zero Lts";
+	compatible = "xunlong,orangepi-zero-lts", "allwinner,sun8i-h2-plus";
+
+};
--- a/configs/orangepi_zero_lts_defconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/configs/orangepi_zero_lts_defconfig	2019-05-03 17:45:39.000000000 +0300
@@ -0,0 +1,17 @@
+CONFIG_ARM=y
+CONFIG_ARCH_SUNXI=y
+CONFIG_SPL=y
+CONFIG_MACH_SUN8I_H3=y
+CONFIG_DRAM_CLK=624
+CONFIG_DRAM_ZQ=3881979
+CONFIG_DRAM_ODT_EN=y
+CONFIG_VIDEO_DE2=y
+CONFIG_SPL_SPI_SUNXI=y
+CONFIG_NR_DRAM_BANKS=1
+# CONFIG_SYS_MALLOC_CLEAR_ON_INIT is not set
+CONFIG_CONSOLE_MUX=y
+# CONFIG_CMD_FLASH is not set
+CONFIG_DEFAULT_DEVICE_TREE="sun8i-h2-plus-orangepi-zero-lts"
+CONFIG_SUN8I_EMAC=y
+CONFIG_USB_EHCI_HCD=y
+CONFIG_SYS_USB_EVENT_POLL_VIA_INT_QUEUE=y
--- a/board/sunxi/MAINTAINERS	2019-01-15 00:02:36.000000000 +0200
+++ b/board/sunxi/MAINTAINERS	2019-05-03 17:50:22.000000000 +0300
@@ -355,6 +355,12 @@
 S:	Maintained
 F:	configs/orangepi_zero_defconfig
 
+ORANGEPI ZERO LTS BOARD
+M:	Icenowy Zheng <icenowy@aosc.xyz>
+M:	Semen Kreyda <aswerklon@gmail.com>
+S:	Maintained
+F:	configs/orangepi_zero_lts_defconfig
+
 ORANGEPI ZERO PLUS BOARD
 M:	Hauke Mehrtens <hauke@hauke-m.de>
 S:	Maintained
