--- a/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
+++ b/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
@@ -161,6 +200,12 @@
 	allwinner,audio-routing = "Line Out", "LINEOUT", 
 				"MIC1", "Mic", "Mic", "MBIAS";
 };
+
+&ir {
+	pinctrl-names = "default";
+	pinctrl-0 = <&ir_pins_a>;
+	status = "disabled";
+};
 
 &spi0 {
 	/* Disable SPI NOR by default: it optional on Orange Pi Zero boards */
