--- a/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
+++ b/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
@@ -189,7 +250,19 @@
 		function = "gpio_in";	// in (initially)
 		pull = <0x00>;		// off
 	};
+
+	pwm0_pin: pwm0 {
+		pins = "PA5";
+		function = "pwm0";
+		drive = <SUN4I_PINCTRL_10_MA>;
+		pull = <SUN4I_PINCTRL_NO_PULL>;
+	};
 };
+
+&pwm {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pwm0_pin>;
+};
 
 &spi0 {
 	/* Disable SPI NOR by default: it optional on Orange Pi Zero boards */
